src/js/basis/constrained_basis.ts

```diff
@@ -1,89 +1,79 @@
+        const { constraints } = config;
+        this._constraints = AtomicConstraints.fromObjects(constraints || []); // `constraints` is an Array with ids
     }
 
-    get constraints() {
-        return this._constraints.array;
+    static fromElementsCoordinatesAndConstraints(
+        config: ElementsCoordinatesAndConstraintsConfig,
+    ): ConstrainedBasis {
+        const basisConfig = this._convertValuesToConfig(config);
+        const constraints = AtomicConstraints.fromValues(config.constraints);
+        return new this({
+            ...basisConfig,
+            constraints: constraints.toJSON() as AtomicConstraintsSchema,
+        });
     }
 
-    set constraints(newConstraints: ConstraintValue[]) {
-        this._constraints = new ArrayWithIds(newConstraints);
+    get constraints() {
+        return this._constraints.toJSON() as AtomicConstraintsSchema;
     }
 
-    getConstraintAsArray() {
-        return this._constraints;
+    set constraints(constraints: AtomicConstraintsSchema) {
+        this._constraints = AtomicConstraints.fromObjects(constraints);
     }
 
     get AtomicConstraints() {
-        return AtomicConstraints.fromArray(this.constraints);
+        return AtomicConstraints.fromObjects(this.constraints);
     }
 
-    /**
-     * Serialize class instance to JSON.
-     * @example As below:
-         {
-            ...Basis.toJSON(),
-            "constraints": [
-                {
-                    "id" : 0,
-                    "value" : [
-                        1,
-                        1,
-                        1
-                    ]
-                },
-            ]
-         }
-     */
-    toJSON(): ConstrainedBasisJSON {
+    override toJSON(): ConstrainedBasisConfig {
         return {
             ...super.toJSON(),
-            constraints: this._constraints.toJSON(),
+            constraints: this.constraints,
         };
     }
 
-    getConstraintByIndex(idx: number): ConstraintValue {
-        return this._constraints.getArrayElementByIndex(idx) || [];
+    getConstraintByIndex(idx: number): AtomicConstraintValue {
+        return this._constraints.getElementValueByIndex(idx) || [false, false, false];
     }
 
     /**
      * Helper function returning a nested array with [element, coordinates, constraints] as elements
      */
-    get elementsCoordinatesConstraintsArray(): [string, Coordinate, ConstraintValue, string][] {
-        return this._elements.array.map((element, idx) => {
-            const coordinates = this.getCoordinateByIndex(idx);
-            const constraints = this.getConstraintByIndex(idx);
+    get elementsCoordinatesConstraintsArray(): [
+        string,
+        Coordinate,
+        AtomicConstraintValue,
+        string,
+    ][] {
+        return this._elements.values.map((element: any, idx: number) => {
+            const coordinate = this.getCoordinateByIndex(idx);
```

Comment:
Nit: to be consistent, let's define `Element` and Label as helper types in the corresponding files also, instead of using 'string'